wagtail/core/actions/copy_for_translation.py

```diff
@@ -0,0 +1,152 @@
+from django.core.exceptions import PermissionDenied
+from django.db import transaction
+
+from wagtail.core.utils import find_available_slug
+
+
+class ParentNotTranslatedError(Exception):
+    """
+    Raised when a call to Page.copy_for_translation is made but the
+    parent page is not translated and copy_parents is False.
+    """
+
+    pass
+
+
+class CopyForTranslationPermissionError(PermissionDenied):
+    """
+    Raised when the page translation copy cannot be performed due to insufficient permissions.
+    """
+
+    pass
+
+
+class CopyForTranslationAction:
+    """
+    Creates a copy of this page in the specified locale.
+
+    The new page will be created in draft as a child of this page's translated
+    parent.
+
+    For example, if you are translating a blog post from English into French,
+    this method will look for the French version of the blog index and create
+    the French translation of the blog post under that.
+
+    If this page's parent is not translated into the locale, then a ``ParentNotTranslatedError``
+    is raised. You can circumvent this error by passing ``copy_parents=True`` which
+    copies any parents that are not translated yet.
+
+    The ``exclude_fields`` parameter can be used to set any fields to a blank value
+    in the copy.
+
+    Note that this method calls the ``.copy()`` method internally so any fields that
+    are excluded in ``.exclude_fields_in_copy`` will be excluded from the translation.
+    """
+
+    def __init__(
+        self,
+        page,
+        locale,
+        copy_parents=False,
+        alias=False,
+        exclude_fields=None,
+        user=None,
+        include_subtree=False,
+    ):
+        self.page = page
+        self.locale = locale
+        self.copy_parents = copy_parents
+        self.alias = alias
+        self.exclude_fields = exclude_fields
+        self.user = user
+        self.include_subtree = include_subtree
+
+    def check(self, skip_permission_checks=False):
+        # Permission checks
+        if (
+            self.user
+            and not skip_permission_checks
+            and not self.user.has_perms(["simple_translation.submit_translation"])
+        ):
+            raise CopyForTranslationPermissionError(
+                "You do not have permission to submit a translation for this page."
+            )
+
+    def walk(self, current_page):
+        for child_page in current_page.get_children():
+            action = CopyForTranslationAction(page=child_page, locale=self.locale)
+            action.execute(skip_permission_checks=True)
+
+            if child_page.numchild:
```

Comment:
I think this optimisation should already exist in ``get_children`` (if it doesn't exist there, we probably shouldn't second-guess the decision to exclude this optimisation here).

I know this is in the original implementation, sorry I missed it when I first reviewed it!

Reply:
I don't understand what you mean. I don't know which optimisation you're talking about.

Should we only remove the line: `if child_page.numchild` or the entire condition block?

Reply:
Yep, I'm saying replace:

```python
if child_page.numchild:
    self.walk(child_page)
```

with:

```python
self.walk(child_page)
```

It looks like this check was implemented to avoid a call to ``page.get_children()`` to stop it running an unnecessary database query (if ``numchild`` is 0, ``get_children()`` would always return an empty list).

But I think this optimisation belongs in the ``get_children`` method itself rather than here.

Reply:
Here it is: https://github.com/django-treebeard/django-treebeard/blob/master/treebeard/mp_tree.py#L1000 (``self.is_leaf()`` checks ``numchild``)

Reply:
Ah I understand better now. I've addressed it, thanks!